src/lib/core/prbuf.c

```diff
@@ -0,0 +1,359 @@
+static bool
+prbuf_is_empty(struct prbuf *buf)
+{
+	return buf->header->end == 0;
+}
+
+/**
+ * Assuming @a record pointing to the valid buffer position, move it to the
+ * next entry. In case of current entry is broken (contains wrong size or
+ * points out of the buffer) function returns -1.
+ */
+static int
+prbuf_record_next(struct prbuf *buf, struct prbuf_record **record)
+{
+	struct prbuf_record *current = *record;
+	*record = NULL;
+	if (current->size > prbuf_size(buf))
+		return -1;
+	if ((char *)current < prbuf_linear_begin(buf))
+		return -1;
+	if ((char *)current > prbuf_linear_end(buf))
+		return -1;
+	char *next_record_ptr = current->data + current->size;
+	if (next_record_ptr > prbuf_linear_end(buf))
+		return -1;
+
+	/* We have reached the end of buffer. */
+	if (next_record_ptr == prbuf_current_raw(buf))
+		return 0;
+
+	current = (struct prbuf_record *)next_record_ptr;
+	if ((uint32_t)(prbuf_linear_end(buf) - next_record_ptr) <
+	    record_size_overhead)
+		current = (struct prbuf_record *)prbuf_linear_begin(buf);
+	else if (current->size == prbuf_end_position)
+		current = (struct prbuf_record *)prbuf_linear_begin(buf);
+	*record = current;
+	return 0;
+}
+
+/**
+ * Verify that prbuf remains in the consistent state: header is valid and
+ * all entries have readable sizes.
+ */
+static bool
+prbuf_check(struct prbuf *buf)
+{
+	if (buf->header->version != prbuf_version)
+		return false;
+	if (buf->header->begin > prbuf_size(buf))
+		return false;
+	if (buf->header->end > prbuf_size(buf))
+		return false;
+	struct prbuf_record *current;
+	current = prbuf_first_record(buf);
+	uint32_t total_size = current->size;
+	int rc = 0;
+	while ((rc = prbuf_record_next(buf, &current)) == 0 &&
+	       current != NULL)
+		total_size += current->size;
+	return rc == 0 && (total_size <= buf->header->size);
+}
+
+int
+prbuf_open(struct prbuf *buf, void *mem)
+{
+	buf->header = (struct prbuf_header *)mem;
+	if (!prbuf_check(buf))
+		return -1;
+	return 0;
+}
+
+/**
+ * Starting from @a current issue skip @a to_store bytes and return
+ * the next record after that.
+ */
+static struct prbuf_record *
+prbuf_skip_record(struct prbuf *buf, struct prbuf_record *current,
+		  ssize_t to_store)
+{
+	assert(to_store > 0);
+	assert(to_store <= buf->header->size);
+
+	while (to_store > 0) {
+		assert(current->size != prbuf_end_position);
+		assert(current->size != 0);
+		to_store -= prbuf_record_alloc_size(current->size);
+		int rc = prbuf_record_next(buf, &current);
+		assert(rc == 0);
+		(void)rc;
+	}
+	return current;
+}
+
+/** Place special mark at the end of buffer to avoid out-of-bound access. */
+static void
+prbuf_set_end_position(struct prbuf *buf)
+{
+	if (prbuf_has_before_end(buf, record_size_overhead))
+		prbuf_current_record(buf)->size = prbuf_end_position;
+}
+
+/** Store entry's size. */
+static void *
+prbuf_prepare_record(struct prbuf_record *record, size_t size)
+{
+	record->size = size;
```

Comment:
AFAIU this must be done on commit, not on prepare.

Please add one more test case:
1. Prepare a record without committing it.
2. Iterate over all records.

The iteration shouldn't include the prepared record, nor the records purged by prepare.

Reply:
I guess no: that size is used in commit. We have to store the size somewhere because `commit` doesn't take the size as an argument; and this is the perfect place to store it.

Reply:
IMO the perfect place to store the size would be `struct prbuf` (the one that isn't persisted; allocated on stack).

Reply:
Added two test cases: see test_buffer_prepared() and test_buffer_prepared_large().
Also added following fixes:
In `prbuf_skip_record()` added NULL check since in corner case we may reach the end of the buffer (see `prepared_large test()` case);
In `prbuf_prepare`:
 - correct size of allocation passed to `prbuf_skip_record` should be `alloc_size - free_space` (since space between head and first record is accounted as well);
 - for the same reason as I've patched `prbuf_skip_record()` I also have to move `end` position in case entry does not fit till the end of the buffer (`prbuf_prepare`) - to the linear begin (`buf->header->end = 0;`). Otherwise we won't be able to say that the buffer is empty (case `prepared_large()`).